res/controllers/lib/mididispatcher.mjs

```diff
@@ -0,0 +1,26 @@
+export class MidiDispatcher {
+    constructor(noteOff) {
+        this.noteOff = noteOff;
+        this.inputMap = new Map();
+    }
+    registerInputCallback(midiBytes, callback) {
+        // JavaScript is broken and believes [1,2] === [1,2] is false, so
+        // JSONify the Array to make it usable as a Map key.
+        const key = JSON.stringify(midiBytes);
+        this.inputMap.set(key, callback);
+        // If passed a Note On message, also map the corresponding Note Off
+        // message to the same callback.
+        if (this.noteOff === true && ((midiBytes[0] & 0xF0) === 0x90)) {
+            const noteOffBytes = [midiBytes[0] - 0x10, midiBytes[1]];
+            const noteOffKey = JSON.stringify(noteOffBytes);
+            this.inputMap.set(noteOffKey, callback);
+        }
+    }
+    receiveData(data, timestamp) {
+        const key = JSON.stringify([data[0], data[1]]);
+        const callback = this.inputMap.get(key);
+        if (typeof callback === 'function') {
```

Comment:
Now that we're using QJSEngine, should we start using asserts to point our programming errors in JS code?

    console.assert(['undefined', 'null', 'function'].includes(typeof callback), "Invalid callback registered")

Support for the console extension can be enabled like this: https://doc.qt.io/qt-5/qjsengine.html#installExtensions

Reply:
or `throw`?

Reply:
I'm in favor of using `throw`, though we should still install the console extension since it enables makes the environment a bit more nodejs-like and also provides additional features such as basic profiling (using `console.time(-End)`) and printing a stacktrace (`console.trace`).

Reply:
It turns out you need to do `throw new Error("message");` for this to work. `throw "message";` does not work.

Reply:
Do we really need to check the function here? We already checked it in `setInputCallback` and introspection and a string comparision on every midi message seems overkill here.

Reply:
`null` is a valid value for the callback. We also need to check for `undefined`. To avoid string comparison, we could explicitly compare to those, but this assumes that the script doesn't inappropriately modify the `inputMap` directly because JS has no private class members.

Reply:
In case the user does inappropriately modify what should be a private member if private members existed in JS, they would get a TypeError thrown when invoking `callback` as a function below. I think that is okay. The script developer should be able to figure out the problem from that.